src/modal-manager.ts

```diff
@@ -0,0 +1,226 @@
+import {
+  LitElement,
+  html,
+  css,
+  customElement,
+  CSSResult,
+  TemplateResult,
+  property,
+  query,
+  PropertyValues,
+} from 'lit-element';
+
+import './modal-template';
+import { ModalTemplate } from './modal-template';
+import { ModalConfig } from './modal-config';
+
+/**
+ * Various modes the modal can be in
+ *
+ * @export
+ * @enum {number}
+ */
+export enum ModalManagerMode {
+  Modal = 'modal',
+  Closed = 'closed',
+}
+
+export interface ModalManagerInterface extends LitElement {
+  /**
+   * Get the current modal mode.
+   */
+  getMode(): ModalManagerMode;
+
+  /**
+   * Show a modal from a given ModalConfig
+   *
+   * @param config ModalConfig
+   * @param customModalContent TemplateResult | undefined
+   * @param userClosedModalCallback () => void | undefined an optional callback when the modal is closed
+   */
+  showModal(options: {
+    config: ModalConfig;
+    customModalContent?: TemplateResult;
+    userClosedModalCallback?: () => void;
```

Comment:
Why are we using a callback instead of emitting a custom event on close? That should prevent the infinite loop as well.

Reply:
This is a momentary action. We show the modal, then some action happens on it that will eventually result in it closing.

To me, event listeners are for long-lived things. Like internally, the `modal-manager-template` emits custom events that the `modal-manager` listens for.

I think using event listeners for something like this, the consumer will end up having to deal with making sure they bind and unbind every time the various actions happen so you don't respond to events that you didn't instantiate.